<a href="https://colab.research.google.com/github/eyesimk/CS412-MachineLearning/blob/main/hw2_kebabci_eceyesim.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# 1) Initialization

First, make a copy of this notebook in your drive

In [ ]:
# Mount to your drive, in this way you can reach files that are in your drive
# Run this cell
# Go through the link that will be showed below
# Select your google drive account and copy authorization code and paste here in output and press enter
# You can also follow the steps from that link
# https://medium.com/ml-book/simplest-way-to-open-files-from-google-drive-in-google-colab-fae14810674 

from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# 2) Loading Dataset

To start working for your homework, take a copy of the folder, given in the below link to your own google drive. You find the train and test data under this folder.

[https://drive.google.com/drive/folders/1DbW6VxLKZv2oqFn9SwxAnVadmn1_nPXi?usp=sharing](https://drive.google.com/drive/folders/1DbW6VxLKZv2oqFn9SwxAnVadmn1_nPXi?usp=sharing)

After copy the folder, copy the path of the train and test dataset to paste them in the below cell to load your data.


In [ ]:
import pandas as pd

train_df = pd.read_csv('/content/drive/My Drive/CS412-HW2/Copy of german_credit_train.csv')
test_df = pd.read_csv('/content/drive/My Drive/CS412-HW2/Copy of german_credit_test.csv')

# 3) Analyze of the Dataset 



In [ ]:
# Print shape
print("Train data dimensionality: ", train_df.shape )

# Print random 5 rows
print("Examples from train data: ")
train_df.head()


Train data dimensionality:  (800, 13)
Examples from train data: 


AccountStatus  Duration CreditHistory  ...  OtherInstallPlans Housing Risk
0           A14        12           A32  ...               A143    A152    1
1           A11         9           A32  ...               A143    A152    1
2           A11        18           A34  ...               A143    A153    1
3           A11        14           A32  ...               A143    A152    2
4           A14        15           A32  ...               A143    A152    2

[5 rows x 13 columns]

In [ ]:
# Print the information about the dataset
print("Information about train data ", train_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   AccountStatus      800 non-null    object
 1   Duration           800 non-null    int64 
 2   CreditHistory      800 non-null    object
 3   CreditAmount       800 non-null    int64 
 4   SavingsAccount     800 non-null    object
 5   EmploymentSince    800 non-null    object
 6   PercentOfIncome    800 non-null    int64 
 7   PersonalStatus     800 non-null    object
 8   Property           800 non-null    object
 9   Age                800 non-null    int64 
 10  OtherInstallPlans  800 non-null    object
 11  Housing            720 non-null    object
 12  Risk               800 non-null    int64 
dtypes: int64(5), object(8)
memory usage: 81.4+ KB
Information about train data  None


# 4) Defining train and test labels




In [ ]:
# Define labels
train_label = train_df['Risk']
test_label = test_df['Risk']

train_df.pop('Risk')
test_df.pop('Risk')

train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   AccountStatus      800 non-null    object
 1   Duration           800 non-null    int64 
 2   CreditHistory      800 non-null    object
 3   CreditAmount       800 non-null    int64 
 4   SavingsAccount     800 non-null    object
 5   EmploymentSince    800 non-null    object
 6   PercentOfIncome    800 non-null    int64 
 7   PersonalStatus     800 non-null    object
 8   Property           800 non-null    object
 9   Age                800 non-null    int64 
 10  OtherInstallPlans  800 non-null    object
 11  Housing            720 non-null    object
dtypes: int64(4), object(8)
memory usage: 75.1+ KB


# 5) Handling missing values




In [ ]:
# Print columns with NaN values
print("NaN value percentage in Housing: %", train_df['Housing'].isnull().sum() /len(train_df) * 100)
#Since it is only %10, we can impute it.

NaN value percentage in Housing: % 10.0


In [ ]:
# Impute missing values by replacing with mode value
train_df['Housing'] = train_df['Housing'].fillna(train_df['Housing'].mode()[0])

test_df['Housing'] = test_df['Housing'].fillna(train_df['Housing'].mode()[0])

# 6) Transforming categorical / ordinal features



In [ ]:
# Transform the categorical / ordinal attributes

#Transformation for ordinal attributes

savings_map = {'A65':0, 'A61':1,'A62':2,'A63':3, 'A64': 4}
train_df['SavingsAccount'] = train_df['SavingsAccount'].replace(savings_map)
test_df['SavingsAccount'] = test_df['SavingsAccount'].replace(savings_map)

accountst_map = {'A14':0, 'A11':1,'A12':2,'A13':3}
train_df['AccountStatus'] = train_df['AccountStatus'].replace(accountst_map)
test_df['AccountStatus'] = test_df['AccountStatus'].replace(accountst_map)

employment_map = {'A71':0,'A72':1,'A73':2,'A74':3,'A75':4}
train_df['EmploymentSince'] = train_df['EmploymentSince'].replace(employment_map)
test_df['EmploymentSince'] = test_df['EmploymentSince'].replace(employment_map)

In [ ]:
#Transformation for categorical attributes
'''
credit_dummies = pd.get_dummies(train_df['CreditHistory'], prefix = 'credit')
print(credit_dummies.head())

personal_dummies = pd.get_dummies(train_df['PersonalStatus'], prefix = 'personal')
print(personal_dummies.head())

housing_dummies = pd.get_dummies(train_df['Housing'], prefix = 'housing')
print(housing_dummies.head())

property_dummies = pd.get_dummies(train_df['Property'], prefix = 'property')
print(property_dummies.head())

otherinstall_dummies = pd.get_dummies(train_df['OtherInstallPlans'], prefix = 'otherinstall')
print(otherinstall_dummies.head())
'''


from sklearn.preprocessing import OneHotEncoder

enc = OneHotEncoder(handle_unknown='ignore')

dummies = enc.fit_transform(train_df[['CreditHistory', 'PersonalStatus', 'Housing','Property','OtherInstallPlans']]).toarray()
dummies = pd.DataFrame(dummies)
train_df = pd.merge(train_df,dummies,right_index = True, left_index = True)
train_df = train_df.drop(columns = ['CreditHistory', 'PersonalStatus', 'Housing','Property','OtherInstallPlans'])

dummies_test = enc.transform(test_df[['CreditHistory', 'PersonalStatus', 'Housing','Property','OtherInstallPlans']]).toarray()
dummies_test = pd.DataFrame(dummies_test)
test_df = pd.merge(test_df,dummies_test,right_index = True, left_index = True)
test_df = test_df.drop(columns = ['CreditHistory', 'PersonalStatus', 'Housing','Property','OtherInstallPlans'])

train_df.head()

AccountStatus  Duration  CreditAmount  SavingsAccount  ...   15   16   17   18
0              0        12          2859               0  ...  1.0  0.0  0.0  1.0
1              1         9          2136               1  ...  0.0  0.0  0.0  1.0
2              1        18          5302               1  ...  1.0  0.0  0.0  1.0
3              1        14          8978               1  ...  0.0  0.0  0.0  1.0
4              0        15          4623               2  ...  0.0  0.0  0.0  1.0

[5 rows x 26 columns]

# 7) Build a k-NN classifier on training data and perform models selection using 5 fold cross validation

*  Initialization of k-NN classifiers with **k= 5, 10, 15**
*  Calculating the cross validation scores using cross_al_score method, number of folds is 5. 
*  Storing the average accuracies of these folds
*  Selecting the value of k using the cross validation results. 

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from statistics import mean
import numpy as np

# k values
kVals = [5,10,15]

# Save the accuracies of each value of kVal in [accuracies] variable
accuracies = []

# Loop over values of k for the k-Nearest Neighbor classifier
for k in kVals:
  # Initialize a k-NN classifier with k neighbors
  neigh = KNeighborsClassifier(n_neighbors= k)
  std = []
  # Calculate the 5 fold cross validation scores using cross_val_score
  # cv parameter: number of folds, in our case it must be 5
  scores = cross_val_score(neigh,train_df,train_label, cv = 5)
  std.append(scores)
  # Stores the average accuracies of the scores in accuracies variable, you can use mean method
  accuracies.append(mean(scores))
  print("Standart Deviation for value:" ,k," ", np.std(std))
print(accuracies)


Standart Deviation for value: 5   0.03292984968079871
Standart Deviation for value: 10   0.006123724356957952
Standart Deviation for value: 15   0.011592023119369639
[0.67625, 0.7075, 0.71]


# 8) Retrain using all training data and test on test set

* Training a classifier with the chosen k value of the best classifier using **all training data**. 

* Predicting the labels of testing data 

* Reporting the accuracy 

In [ ]:
from sklearn.metrics import accuracy_score


# Train the best classifier using all training set
best_kNN = KNeighborsClassifier(n_neighbors= kVals[np.argmax(accuracies)])
best_kNN.fit(train_df, train_label)

# Estimate the prediction of the test data
preds = best_kNN.predict(test_df)

# Print accuracy of test data
accuracy = accuracy_score(test_label, preds)
print(accuracy)

0.665


In [ ]:
from sklearn import preprocessing
normalized_train = preprocessing.normalize(train_df)
normalized_test = preprocessing.normalize(test_df)

# Train the best classifier using all training set
best_kNN = KNeighborsClassifier(n_neighbors= 15)
best_kNN.fit(normalized_train, train_label)

# Estimate the prediction of the test data
preds = best_kNN.predict(normalized_test)

# Print accuracy of test data
accuracy = accuracy_score(test_label, preds)
print(accuracy)

0.7


In [ ]:
#I obtained a better test accuracy with k = 15, feature normalization within the rate of 70%. 

# 9) Report


In this assignment, the aim was to classify people as bad or good credit risks according to the given attributes from german credit dataset by building a k-NN classifier. Before starting, I made some operations for preprecessing. Firstly, with the shape() function I observed how many entries exists in the dataset, later within the info() function I realised that there are missing values for Housing column.When I calculated the percentage of missing values, I have reached the solution 10%, so I decided to impute it. Afterward I transformed the Ordinal features(Account Status, Account Savings, Employment Since) with mapping and the Categorical features (Creedit History, Personal Status, Property, Other Install Plans, Housing) with one hot encoding.After the preprocessing is over, I started building k-NN classifiers with k= 5, 10, 15. The average validation accuracies were: 
*  k= 5 -> 0.67625 
*  k= 10 -> 0.7075
*  k= 15 -> 0.71 

And the corresponding standart deviations were:  

*   k= 5 -> 0.03292984968079871
*   k= 10 -> 0.006123724356957952
*   k= 15 -> 0.011592023119369639

According to the validation accuracies I have chosen the k value = 15, with accuracy score 71%. Afterward, I have obtained 66.5% accuracy rate on test data within the chosen classifier (k = 15).
For the bonus part, I decided to do feature normalization and choose the same value for kNN classifier.When the necessary operations were completed I obtained the classification accuracy of 70% on test data with (k=15) which is  better than the first result.

 